x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectProcessManager.java

```diff
@@ -458,8 +458,11 @@ protected void doRun() {
 
                                     try {
                                         createProcessAndSetRunning(processContext, job, params, closeHandler);
-                                        processContext.getAutodetectCommunicator().init(params.modelSnapshot());
+                                        processContext.getAutodetectCommunicator().restoreState(params.modelSnapshot());
                                         setJobState(jobTask, JobState.OPENED);
+
+                                        assert processByAllocation.get(jobTask.getAllocationId()) == null;
+                                        processByAllocation.put(jobTask.getAllocationId(), processContext);
```

Comment:
The fix for the bug is the `writeHeader` change but I thought moving this here would be a sensible change. 

Once added to the `processByAllocation` map the process can be used so why add `processContext` before it is fully formed? There are work arounds in the close() function for the case where `processContext` does not have an AutodetectCommunicator and I can't see the reason why it was not set here before.

Reply:
Before you made the job in index changes `processByAllocation.put(jobTask.getAllocationId(), processContext);` was called at the very beginning of `openJob` before any async actions were called.  Currently it's in between two async calls which I agree solves nothing.  But there may have originally been a good reason why it was done at the beginning of `openJob` before any async calls were made, for example to fix a race condition when two identical calls to `open_job` are made in very quick succession.  @dimitris-athanasiou can you remember?